syntax = "proto3";

package ServerRpc.group;

import "user.proto";

message groupMember{
    user.user user = 1;
    string groupID = 2;
    enum groupMemberRole{
        owner = 0;
        member = 1;
    }
    groupMemberRole role = 3;
}

message groupInfo{
    string groupName = 1;
    string groupDesc = 2;
    string groupOwnerId = 3;
    string groupId = 4;
    repeated groupMember groupMembers = 5;
}

message createGroupReq {
    string groupName = 1;
    string groupDesc = 2;
    string groupOwnerId = 3;
    string groupId = 4;
    repeated string groupMemberIds = 5;
}

message createGroupResp {
}

message joinGroupReq {
    string groupID = 1;
    string userID = 2;
}

message joinGroupResp {
}

message getGroupInfoReq {
    string groupID = 1;
}

message getGroupInfoResp {
    groupInfo groupInfo = 2;
}

message getGroupListReq {
    string userID = 1;
}

message getGroupListResp {
    repeated groupInfo groupInfo = 2;
}

message getGroupMemberIdReq {
    string groupID = 1;
}

message getGroupMemberIdResp {
    repeated string groupMemberIds = 2;
}

message getGroupMemberIdHashReq {
    string groupID = 1;
}

message getGroupMemberIdHashResp {
    uint64 groupMemberIdHash = 2;
}

service Group {
    rpc createGroup(createGroupReq) returns (createGroupResp);
    rpc joinGroup(joinGroupReq) returns (joinGroupResp);
    rpc getGroupInfo(getGroupInfoReq) returns (getGroupInfoResp);
    rpc getGroupList(getGroupListReq) returns (getGroupListResp);
    rpc getGroupMemberId(getGroupMemberIdReq) returns (getGroupMemberIdResp);
    rpc getGroupMemberIdHash(getGroupMemberIdHashReq) returns (getGroupMemberIdHashResp);
}
